common/namespace/registry.go

```diff
@@ -483,6 +483,200 @@ func (r *registry) getNamespaceByIDLocked(id ID) (*Namespace, error) {
+// to the persistence layer and updates caches if it doesn't
+func (r *registry) getOrPutNamespaceByID(id ID) (*Namespace, error) {
+	// check caches
+	cachedNS, cachedErr := r.checkCachesByID(id)
+	if cachedNS != nil || cachedErr != nil {
+		return cachedNS, cachedErr
+	}
+
+	r.readthroughRequestLock.Lock()
+	defer r.readthroughRequestLock.Unlock()
+
+	// check caches again in case there was an update while waiting
+	cachedNS, cachedErr = r.checkCachesByID(id)
+	if cachedNS != nil || cachedErr != nil {
+		return cachedNS, cachedErr
+	}
+
+	// readthrough to persistence layer and update readthrough cache if not found
+	ns, err := r.getNamespaceByIDPersistence(id)
+	if err != nil {
+		return nil, err
+	}
+
+	// update main caches if found
+	r.updateCachesSingleNamespace(ns)
+
+	return ns, nil
+}
+
+func (r *registry) checkCachesByID(id ID) (*Namespace, error) {
+	// check main caches
+	cacheHit, cacheErr := r.getNamespaceByID(id)
+	if cacheErr == nil {
+		return cacheHit, nil
+	}
+
+	// check readthrough error cache to prevent excessive requests to persistence layer
+	prevErr := r.getPreviousReadthroughError(id.String())
+	if prevErr != nil {
+		return nil, prevErr
+	}
+
+	// missed both caches, should readthrough
+	return nil, nil
+}
+
+func (r *registry) getPreviousReadthroughError(handle string) error {
+	r.readthroughCacheLock.RLock()
+	defer r.readthroughCacheLock.RUnlock()
+
+	if prevErr, ok := r.readthroughErrorsCache.Get(handle).(error); ok {
+		return prevErr
+	}
+	return nil
+}
+
+func (r *registry) updateCachesSingleNamespace(ns *Namespace) {
+	var stateChangeCallbacks []StateChangeCallbackFn
+
+	r.cacheLock.Lock()
+	if curEntry, ok := r.cacheByID.Get(ns.ID()).(*Namespace); ok {
+		if curEntry.ConfigVersion() >= ns.ConfigVersion() {
+			// More up to date version already put in cache by refresh
+			r.cacheLock.Unlock()
+			return
+		}
+	}
+
+	oldNS := r.updateIDToNamespaceCache(r.cacheByID, ns.ID(), ns)
+	r.cacheNameToID.Put(ns.Name(), ns.ID())
+	if namespaceStateChanged(oldNS, ns) {
+		stateChangeCallbacks = mapAnyValues(r.stateChangeCallbacks)
+	}
+	r.cacheLock.Unlock()
+
+	// call state change callbacks
+	for _, cb := range stateChangeCallbacks {
+		cb(ns, false)
+	}
+}
+
+func (r *registry) updateReadthroughCache(identifier string, err error) {
+	r.readthroughCacheLock.Lock()
+	defer r.readthroughCacheLock.Unlock()
+	r.readthroughErrorsCache.Put(identifier, err)
+}
+
+func (r *registry) getNamespaceByNamePersistence(name Name) (*Namespace, error) {
+	request := &persistence.GetNamespaceRequest{
+		Name: name.String(),
+	}
+
+	ns, err := r.getNamespacePersistence(request)
+	if err != nil {
+		notFoundErr := serviceerror.NewNamespaceNotFound(name.String())
+		if _, ok := err.(*serviceerror.NamespaceNotFound); ok {
+			// TODO: we should cache and return the actual error we got (e.g. timeout)
+			r.updateReadthroughCache(name.String(), notFoundErr)
```

Comment:
Since we eventually plan on returning real errors returned by persistence, or other errors like timeouts, I think it makes sense to cache the actual error.